perceval/simulators/simulator.py

```diff
@@ -390,6 +390,32 @@ def probs_svd(self, input_dist: SVDistribution, progress_callback: Optional[Call
                 'physical_perf': self._physical_perf,
                 'logical_perf': self._logical_perf}
 
+    def probs_density_matrix(self, dm: DensityMatrix) -> dict:
+        """
+        gives the output probability distribution, after evolving some density matrix through the simulator
+        :param dm: the input DensityMatrix
+        """
+        if not isinstance(dm, DensityMatrix):
+            raise TypeError(f"dm must be a DensityMatrix object, {type(dm)} was given")
+
+        input_list = self._get_density_matrix_input_list(dm)
+        u_evolve = self._construct_evolve_operator(input_list, dm)
+        u_evolve_in_row = csr_array(u_evolve)
```

Comment:
Simply to clarify, you use `csc_array` for evolve and `csr_array` for probs - any specific reason?

Also, `dok_array` is used in `DensityMatrix` (I see that it seems easier to build from scratch). Perhaps some documentation/comment would be nice to help remember their usage for posterity.

Reply:
csc_arrays can be constructed by columns (and then we can iterate on columns), and that's what I do for the evolve operator because each column correspond to the output of one Fockstate. However in probs, I need to iterate on the rows so I convert the operator into this format first.
dok_array are less efficient but I can add values anywhere in the matrix easily, so that's what I do for very sparse matrix, such as basis change matrix, etc
I added some comments.